json_patch.c

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (c) 2021 Alexandru Ardelean.
+ *
+ * This is free software; you can redistribute it and/or modify
+ * it under the terms of the MIT license. See COPYING for details.
+ *
+ */
+
+#include "config.h"
+
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+
+#include "json_patch.h"
+#include "json_object_private.h"
+
+/**
+ * JavaScript Object Notation (JSON) Patch
+ *   RFC 6902 - https://tools.ietf.org/html/rfc6902
+ */
+
+static int json_patch_apply_test(struct json_object **res,
+                                 struct json_object *patch_elem,
+                                 const char *path)
+{
+	struct json_object *value1, *value2;
+
+	if (!json_object_object_get_ex(patch_elem, "value", &value1)) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	/* errno should be set by json_pointer_get() */
+	if (json_pointer_get(*res, path, &value2))
+		return -1;
+
+	if (!json_object_equal(value1, value2)) {
+		json_object_put(*res);
+		*res = NULL;
+		errno = ENOENT;
+		return -1;
+	}
+
+	return 0;
+}
+
+static int __json_patch_apply_remove(struct json_pointer_get_result *jpres)
+{
+	if (json_object_is_type(jpres->parent, json_type_array)) {
+		return json_object_array_del_idx(jpres->parent, jpres->id.index, 1);
+	} else if (jpres->parent && jpres->id.key) {
+		json_object_object_del(jpres->parent, jpres->id.key);
+		return 0;
+	} else {
+		return json_object_put(jpres->obj);
```

Comment:
Is this the case where you're removing the root object?  If so, doesn't that mean that *res is now invalid, and should be set to NULL?  Is removing the root even a valid thing to do?

Reply:
Fixed at https://github.com/hawicz/json-c/commit/670bf12f9ee5cef46a116c3350b78cf4a358ec2b